File: 01/Jonas/utility/christmas-printer.ts
export interface IChristmasPrinterInput {
    suggestedAnswer: number | string;
    answer: number | string;
}

export class ChristmasPrinter {

    public print(value: IChristmasPrinterInput, print?: boolean): void {
        if (print) {
            console.log('         ┌-------------------------------------------------------------------------------------------------┐');
            console.log('         |                                                 ┌------------------╲                ╲   ╱       |');
            console.log('         |╲            ^                                   | North Pole 807km  >              ─  *  ─      |');
            console.log('         |  ╲        ╱   ╲               ^                 └------------------╱                 ╱o╲        |');
            console.log('         |    ╲    ╱~~~~~~~╲           ╱~~~╲                       |  |                        ╱*~~╲       |');
            console.log('         |      ╲╱           ╲     ╱╲╱       ╲  ┌-----------------------------------┬----┐    ╱o~*~o╲      |');
            console.log(`         |        ╲__     ╱╲   ╲_╱    ╲____    ╲|     ${this.getChristmasCountdown(30)}╰─╮  |   ╱~o~o~~*╲     |`);
            console.log(`         |           ╲__╱    ╲_╱        ╲       | Suggested Answer: ${this.getMessageWithAvailableTrailingSpaces(value.suggestedAnswer, 17)} ╰──|  ╱*~~o~*~~o╲    |`);
            console.log(`         |      ╱╲                        ╲___  | Answer:           ${this.getMessageWithAvailableTrailingSpaces(value.answer, 20)} | ╱~o~~*~o~~*~╲   |`);
            console.log('         |     ╱˷˷╲    ,_____,     ╭╮╭───────╯  └----------------------------------------┘╱o~*~~o~*~~o~~╲  |');
            console.log('         |   ╭──||────/_╲_____╲────╯╰╯                             |  |               ┌─┬─┐`˜`˜|˜˜˜|˜`˜´┌┬┐|');
            console.log('         |   ╰────    |_|_____|                                    |  |               └─┴┬┼┐┌─┬┴┐ ┌┼┐┌─┬┴┤┘|');
            console.log('         |                                                        ~~~~~~                 └┴┘└─┴─┘~└┴┘└─┴─┘ |');
            console.log('         └-------------------------------------------------------------------------------------------------┘');
        }
    }

    private getMessageWithAvailableTrailingSpaces(message: number | string, spaceAvailable: number): string {
        return message.toString().padEnd(spaceAvailable);
    }

    private getChristmasCountdown(spaceAvailable: number): string {
        const daysUntilChristmas: number = 24 - new Date().getDate();
        const message: string = daysUntilChristmas > 0 ? `Days Until Christmas ${daysUntilChristmas}` : 'Merry Christmas!';
        const padStart: number = Math.round((spaceAvailable - message.length) / 2);
        const padEnd: number = spaceAvailable - padStart;
        return ''.padStart(padStart) + message.padEnd(padEnd);
    }
}

File: 2024/09/Jonas/util/christmas_printer/ChristmasPrinter.java
package util.christmas_printer;

import util.string_formatter.Message;
import util.string_formatter.replace_placeholder_formatter.ReplacePlaceholderFormatter;
import util.time.DateOfYear;
import util.time.Time;

import java.time.temporal.ChronoUnit;
import java.util.Comparator;
import java.util.stream.Stream;

public abstract class ChristmasPrinter {

    private AdventOfCodeAnswer adventOfCodeAnswer;
    private final DateOfYear christmasDate;
    private final String defaultAnswer = "?";

    protected ChristmasPrinter(DateOfYear christmasDate) {
        this.adventOfCodeAnswer = new AdventOfCodeAnswer(defaultAnswer, defaultAnswer);
        this.christmasDate = christmasDate;
    }

    public void print(int answer, int expected) {
        setAdventOfCodeAnswer(answer, expected);
        print();
    }

    public void print(int answer) {
        setAdventOfCodeAnswer(answer);
        print();
    }

    public void print(AdventOfCodeAnswer adventOfCodeAnswer) {
        this.adventOfCodeAnswer = adventOfCodeAnswer;
        print();
    }

    public void print() {
        System.out.println(getFormattedDesign());
    }

    private void setAdventOfCodeAnswer(int answer) {
        this.adventOfCodeAnswer = new AdventOfCodeAnswer(String.valueOf(answer), defaultAnswer);
    }

    public void setAdventOfCodeAnswer(int answer, int expected) {
        this.adventOfCodeAnswer = new AdventOfCodeAnswer(String.valueOf(answer), String.valueOf(expected));
    }

    private String getFormattedDesign() {
        Message design = getDesign();
        Stream.of(
                getChristmasCountdownDesignTask(),
                getAnswerDesignTask(),
                getExpectedAnswerDesignTask()
        )
                .sorted(Comparator.comparingInt(ChristmasPrinterDesignTask::order))
                .forEach(designTask -> design.format(designTask.replacePlaceholderFormatter(), designTask.value()));
        return design.toString();
    }

    protected abstract Message getDesign();

    protected abstract ReplacePlaceholderFormatter getChristmasCountdownFormatter();

    protected abstract ReplacePlaceholderFormatter getAnswerFormatter();

    protected abstract ReplacePlaceholderFormatter getExpectedAnswerFormatter();

    protected abstract int getChristmasCountdownOrder();

    protected abstract int getAnswerOrder();

    protected abstract int getExpectedAnswerOrder();

    private ChristmasPrinterDesignTask getChristmasCountdownDesignTask() {
        return new ChristmasPrinterDesignTask(
                getChristmasCountdown(),
                getChristmasCountdownOrder(),
                getChristmasCountdownFormatter()
        );
    }

    private ChristmasPrinterDesignTask getAnswerDesignTask() {
        return new ChristmasPrinterDesignTask(
                adventOfCodeAnswer.answer(),
                getAnswerOrder(),
                getAnswerFormatter()
        );
    }

    private ChristmasPrinterDesignTask getExpectedAnswerDesignTask() {
        return new ChristmasPrinterDesignTask(
                adventOfCodeAnswer.expectedAnswer(),
                getExpectedAnswerOrder(),
                getExpectedAnswerFormatter()
        );
    }

    private String getChristmasCountdown() {
        long daysUntilChristmas = getDaysUntilChristmas();
        return daysUntilChristmas > 0 ? String.format("Days Until Christmas %s", daysUntilChristmas) : "Merry Christmas!";
    }

    private long getDaysUntilChristmas() {
        Time today = Time.now();
        Time christmas = Time.of(today.getYear(), this.christmasDate.month(), this.christmasDate.day());

        if (today.getMonth() == christmas.getMonth() && today.getDayOfMonth() == christmas.getDayOfMonth()) {
            return 0;
        } else if (today.toEpochMilli() > christmas.toEpochMilli()) {
            christmas = christmas.add(1, ChronoUnit.YEARS);
        }

        return today.getDaysUntilTime(christmas);
    }
}
